R/filter_confirmation.R

```diff
@@ -0,0 +1,473 @@
+      select(data, !!!by_vars, !!!join_vars, tmp_obs_nr_filter_confirmation),
+      by = vars2chr(by_vars),
+      suffix = c("", ".join")
+    ) %>%
+    filter(tmp_obs_nr_filter_confirmation <= tmp_obs_nr_filter_confirmation.join)
+
+  if (!quo_is_null(first_cond)) {
+    # select all observations up to the first confirmation observation
+    data_joined <- filter_relative(
+      data_joined,
+      by_vars = vars(!!!by_vars, tmp_obs_nr_filter_confirmation),
+      condition = !!first_cond &
+        tmp_obs_nr_filter_confirmation < tmp_obs_nr_filter_confirmation.join,
+      order = vars(tmp_obs_nr_filter_confirmation.join),
+      mode = "first",
+      selection = "before",
+      inclusive = TRUE,
+      keep_no_ref_groups = FALSE
+    )
+  }
+
+  # apply confirmation condition, which may include summary functions
+  data_joined %>%
+    group_by(!!!by_vars, tmp_obs_nr_filter_confirmation) %>%
+    filter(!!filter) %>%
+    # select one observation of each group, as the joined variables are removed
+    # it doesn't matter which one, so we take just the first one
+    slice(1L) %>%
+    ungroup() %>%
+    select(colnames(dataset))
+}
+
+#' Count Number of Observations Where a Variable Equals a Value
+#'
+#' Count number of observations where a variable equals a value.
+#'
+#' @param var A vector
+#'
+#' @param val A value
+#'
+#' @author Stefan Bundfuss
+#'
+#' @keywords user_utility
+#'
+#' @export
+#'
+#' @examples
+#'
+#' library(tibble)
+#' library(dplyr)
+#' library(admiral)
+#' data <- tibble::tribble(
+#'   ~USUBJID, ~AVISITN, ~AVALC,
+#'   "1",      1,        "PR",
+#'   "1",      2,        "CR",
+#'   "1",      3,        "NE",
+#'   "1",      4,        "CR",
+#'   "1",      5,        "NE",
+#'   "2",      1,        "CR",
+#'   "2",      2,        "PR",
+#'   "2",      3,        "CR",
+#'   "3",      1,        "CR",
+#'   "4",      1,        "CR",
+#'   "4",      2,        "NE",
+#'   "4",      3,        "NE",
+#'   "4",      4,        "CR",
+#'   "4",      5,        "PR"
+#' )
+#'
+#' # add variable providing the number of NEs for each subject
+#' group_by(data, USUBJID) %>%
+#'   mutate(nr_nes = count_vals(var = AVALC, val = "NE"))
+count_vals <- function(var, val) {
+  length(var[var == val])
+}
+
+#' Minimum Value on a Subset
+#'
+#' The function derives the minimum value of a vector/column on a subset of
+#' entries/observations.
+#'
+#' @param var A vector
+#'
+#' @param cond A condition
+#'
+#' @author Stefan Bundfuss
+#'
+#' @keywords user_utility
+#'
+#' @export
+#'
+#' @examples
+#'
+#' library(tibble)
+#' library(dplyr)
+#' library(admiral)
+#' data <- tribble(
+#'   ~USUBJID, ~AVISITN, ~AVALC,
+#'   "1",      1,        "PR",
+#'   "1",      2,        "CR",
+#'   "1",      3,        "NE",
+#'   "1",      4,        "CR",
+#'   "1",      5,        "NE",
+#'   "2",      1,        "CR",
+#'   "2",      2,        "PR",
+#'   "2",      3,        "CR",
+#' )
+#'
+#' # add variable indicating if PR occurred after CR
```

Comment:
Now need to add bit more description to this example such as "In oncology setting, when needing to check the first time a patient had a Complete Response (CR) to compare to see if any Partial Response (PR) occurred after this".